In [1]:
repo url: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>
commit id: 41b7a1f33ed27a819a0e19cdf8c2ad5f87c8877d

SyntaxError: invalid decimal literal (2670578440.py, line 2)

In [1]:
# Imports section
import pandas as pd
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, root_mean_squared_error
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
science_data_large = os.path.join(os.getcwd(), "science_data_large.csv")
df = pd.read_csv(science_data_large)
# Output the first 15 rows of the data
print(df.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df[['Temperature °C', 'Mols KCL']]  #features
y = df['Size nm^3'] #label

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set
model = LinearRegression() # Create the  LinearRegression model object
model.fit(X_train, y_train) # Fit the model to the training data 
prediction = model.predict(X_test)  #  Make predictions on the test data 

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = [[500, 500]]
sample_prediction = model.predict(sample_data)
print(sample_prediction)

[540029.26034545]


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [5]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
r2 = r2_score(y_test, prediction)
print(r2)

0.8552472077276095


The R² score for the model indicates the proportion of the variance in the dependent variable (Size nm³) that is predictable from the independent variables (Temperature °C and Mols KCL). An R² score of 1 indicates a perfect fit, while a score of 0 means that the model does not explain any of the variability of the response data around its mean. The R² score for this score is around 0.855 which indicates that it is quite close to being a good fit.

In [6]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

# Function to round to 5 significant figures
def round_to_significant_figures(num, sig_figs=5):
    if num == 0:
        return 0
    else:
        return round(num, sig_figs - int(np.floor(np.log10(abs(num)))) - 1)
    
rounded_intercept = round_to_significant_figures(intercept)
rounded_coefficients = [round_to_significant_figures(coef) for coef in coefficients]

# Equation in LaTeX format:
# h(x) = intercept + coefficient1 * Temperature + coefficient2 * Mols KCL
print(f"h(x) = {rounded_intercept} + {rounded_coefficients[0]} * Temperature + {rounded_coefficients[1]} * Mols KCL")

h(x) = -409390.0 + 866.15 * Temperature + 1032.7 * Mols KCL


Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [7]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(model, X, y, cv=5) #5 fold

# Report on their finding and their significance
print("Cross-Validation Scores:", scores)
print("Average Cross-Validation Score:", np.mean(scores))
print("Standard Deviation of Scores:", np.std(scores))

Cross-Validation Scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Average Cross-Validation Score: 0.8568167899144437
Standard Deviation of Scores: 0.01346630737209602


## Part 5. Using Polynomial Regression

In [10]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2) # Create Polynomial Features of degree 2
X_poly = poly.fit_transform(X)

# Split the dataset into training and testing sets (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X_poly, y, test_size=0.1, random_state=42)

# Train linear regression model on the polynomial features
poly_model = LinearRegression()
poly_model.fit(X_train, y_train)

# Predictions using the polynomial regression model
poly_prediction = poly_model.predict(X_test)

# Report on the metrics and output the resultant equation as you did in Part 3.
r2 = r2_score(y_test, poly_prediction)
print(r2)


1.0


The R² score for the model indicates the proportion of the variance in the dependent variable (Size nm³) that is predictable from the independent variables (Temperature °C and Mols KCL). An R² score of 1 indicates a perfect fit, while a score of 0 means that the model does not explain any of the variability of the response data around its mean. The R² score for this score is a perfect 1 which indicates that it is quite close to being a good fit.

In [11]:

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = poly_model.coef_
intercept = poly_model.intercept_
    
rounded_intercept = round_to_significant_figures(intercept)
rounded_coefficients = [round_to_significant_figures(coef) for coef in coefficients]

# Equation in LaTeX format:
# h(x) = intercept + coefficient1 * Temperature + coefficient2 * Mols KCL
print(f"h(x) = {rounded_intercept} + {rounded_coefficients[0]} * Temperature + {rounded_coefficients[1]} * Mols KCL")

h(x) = 2.0477e-05 + 0 * Temperature + 12.0 * Mols KCL
